File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    static int[] uf;
    public int find(int a){
        if (uf[a]<0){
            return a;
        }
        a=find(uf[a]);
        return a;
    }
    public void merge(int a,int b){
        int p_a=find(a);
        int p_b=find(b);
        if (p_a==p_b){
            return;
        }
        uf[p_a]+=uf[p_b];
        uf[p_b]=p_a;
    }
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        
        for (int i=0;i<wires.length;i++){
            uf=new int[n+1];
            Arrays.fill(uf,-1);
            for (int j=0;j<wires.length;j++){
                if (i==j){
                    continue;
                }
                merge(wires[j][0],wires[j][1]);
            }
            int[] temp=new int[2];
            int index=0;
            for (int j=1;j<uf.length;j++){
                if(uf[j]<0){
                    temp[index]=uf[j];
                    index++;
                }
            }
            
            answer=Math.min(answer,Math.abs(temp[0]-temp[1]));
        }
        return answer;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = n

    graph = [[] for _ in range(n + 1)]
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)

    def DFS(node, visited):
        count=1
        visited[node]=True
        for n in graph[node]:
            if not visited[n]:
                count+=DFS(n,visited)
        return count

    for a, b in wires:
        graph[a].remove(b)
        graph[b].remove(a)

        visited = [False] * (n + 1)
        count = DFS(a, visited)

        diff = abs((n - count) - count)
        answer = min(answer, diff)

        graph[a].append(b)
        graph[b].append(a)

    return answer
